lib/likelihood.c

```diff
@@ -0,0 +1,193 @@
+** it under the terms of the GNU General Public License as published by
+** the Free Software Foundation, either version 3 of the License, or
+** (at your option) any later version.
+**
+** msprime is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with msprime.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+ * Implementation of log-likelihood functions for full-ARG.
+ */
+
+#include <float.h>
+#include <math.h>
+#include <stdlib.h>
+#include "likelihood.h"
+#include "msprime.h"
+
+static double
+get_total_material(tsk_treeseq_t *ts) {
+    double ret = 0;
+    tsk_size_t j;
+    tsk_edge_table_t *edges = &ts->tables->edges;
+    tsk_node_table_t *nodes = &ts->tables->nodes;
+    tsk_id_t parent;
+    tsk_id_t child;
+    
+    for (j = 0; j < edges->num_rows; j++) {
+        parent = edges->parent[j];
+        child = edges->child[j];
+        ret += (edges->right[j] - edges->left[j])
+            * (nodes->time[parent] - nodes->time[child]);
+    }
+    return ret;
+}
+
+int
+msp_unnormalised_log_likelihood_mut(tsk_treeseq_t *ts, double theta,
+                                    double *r_lik) {
+    int ret = 1;
+    tsk_size_t j;
+    tsk_mutation_t mut;
+    tsk_size_t num_mutations = tsk_treeseq_get_num_mutations(ts);
+    const double total_material = get_total_material(ts);
+    const double *node_time = ts->tables->nodes.time;
+    double branch_length, lik;
+    tsk_tree_t tree;
+    tsk_id_t parent, child, it;
+    
+    ret = tsk_tree_init(&tree, ts, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    if (theta > 0) {
+        lik = num_mutations * log(total_material * theta) -
+            total_material * theta;
+        for (it = tsk_tree_first(&tree); it == 1; it = tsk_tree_next(&tree)) {
+            for (j = 0; j < tree.sites_length; j++) {
+                if (tree.sites[j].mutations_length != 1) {
+                    ret = MSP_ERR_BAD_PARAM_VALUE;
+                    goto out;
+                }
+                mut = tree.sites[j].mutations[0];
+                child = mut.node;
+                parent = tree.parent[child];
+                branch_length = node_time[parent] - node_time[child];
+                child = parent;
+                parent = tree.parent[parent];
+                while (parent != TSK_NULL && tree.left_child[child]
+                        == tree.right_child[child]) {
+                    branch_length += node_time[parent] - node_time[child];
+                    child = parent;
+                    parent = tree.parent[parent];
+                }
+                child = mut.node;
+                while (tree.left_child[child] >= 0 && tree.left_child[child]
```

Comment:
I'm not following this loop; it looks like you're iterating over the siblings of ``child``, but I think that's what it's doing. Can you explain what you're doing here?

Reply:
This iterates leafwards from the mutation, and checks how much branch length there is in which the mutation could have taken place without changing the pattern of observed SNPs. There are two ways in which that window can end: either we reach a leaf, or there is a bifurcation, in which case the mutation cannot have happened any further leafwards. 

`tree.left_child[child] >= 0` catches when we reach a leaf, and `tree.left_child[child] == tree.right_child[child]` catches when a bifurcation happens, i.e. when the valid branch length for the given mutation ends. Unary nodes have `tree.left_child[node] == tree.right_child[node]`.

Reply:
I see, this is subtle and clever. Can you put this in as a comment please?

It would be a bit more idiomatic to catch the leaf case with ``tree.left_child[child]  != TSK_NULL``.